viam-cartographer/src/carto_facade_go/carto_facade.go

```diff
@@ -0,0 +1,278 @@
+package cartofacade
+
+/*
+	#cgo CFLAGS: -I../carto_facade
+  	#cgo LDFLAGS: -v -L../../build  -lviam-cartographer -lglog -lstdc++ -lpthread
+
+	#include "../carto_facade/carto_facade.h"
+	// TODO: confirm bstring import is necessary and why
+	#include "bstrlib.h"
+
+	viam_carto_lib* alloc_viam_carto_lib() { return (viam_carto_lib*) malloc(sizeof(viam_carto_lib)); }
+	void free_alloc_viam_carto_lib(viam_carto_lib* p) { free(p); }
+
+	viam_carto* alloc_viam_carto() { return (viam_carto*) malloc(sizeof(viam_carto)); }
+	void free_alloc_viam_carto(viam_carto* p) { free(p); }
+
+	bstring* alloc_bstring_array(size_t len) { return (bstring*) malloc(len * sizeof(bstring)); }
+	void free_bstring_array(bstring* p) { free(p); }
+*/
+import "C"
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"unsafe"
+
+	"github.com/golang/geo/r3"
+	"go.viam.com/rdk/spatialmath"
+)
+
+func successfulStatus(status C.int) bool {
+	// return int(status) == int(C.VIAM_CARTO_SUCCESS)
+	return true
+}
+
+// CViamCartoLib is a struct to hold the c type viam_carto_lib
+type CViamCartoLib struct {
+	viamCartolib *C.viam_carto_lib
+}
+
+// NewViamCartoLib calls viam_carto_lib_init and returns a pointer to a viam carto lib object.
+func NewViamCartoLib(miniloglevel, verbose int) (CViamCartoLib, error) {
+	pVcl := C.alloc_viam_carto_lib()
+	defer C.free_alloc_viam_carto_lib(pVcl)
+
+	ppVcl := (**C.viam_carto_lib)(unsafe.Pointer(pVcl))
+
+	status := C.viam_carto_lib_init(ppVcl, C.int(miniloglevel), C.int(verbose))
+	if !successfulStatus(status) {
+		// TODO: how to make this error useful?
+		return CViamCartoLib{}, fmt.Errorf("error initializing viam_carto_lib status = %d", status)
+	}
+
+	vcl := CViamCartoLib{viamCartolib: (*ppVcl)}
+
+	return vcl, nil
+}
+
+// TerminateViamCartoLib calls viam_carto_lib_terminate to clean up memory for viam carto lib.
+func (vcl *CViamCartoLib) TerminateViamCartoLib() error {
+	status := C.viam_carto_lib_terminate((**C.viam_carto_lib)(unsafe.Pointer(vcl)))
+	if !successfulStatus(status) {
+		// TODO: how to make this error useful?
+		return errors.New(string(status))
+	}
+	return nil
+}
+
+// CViamCarto is a struct to hold the c type viam_carto
+type CViamCarto struct {
+	viamCarto *C.viam_carto
+}
+
+// TODO: properly pass config
+func getConfig() C.viam_carto_config {
+	vcc := C.viam_carto_config{}
+
+	sz := 1
+	pSensor := C.alloc_bstring_array(C.size_t(sz))
+	sensorSlice := unsafe.Slice(pSensor, sz)
+	sensorSlice[0] = C.bfromcstr(C.CString("rplidar"))
+
+	vcc.sensors = pSensor
+	vcc.sensors_len = C.int(sz)
+
+	dataDir := C.bfromcstr(C.CString("data_dir"))
+	vcc.data_dir = dataDir
+
+	componentReference := C.bfromcstr(C.CString("comp_ref"))
+	vcc.component_reference = componentReference
+
+	return vcc
+}
+
+func cleanConfig(vcc C.viam_carto_config) {
+	C.free_bstring_array(vcc.sensors)
+}
+
+// NewViamCarto calls viam_carto_init and returns a pointer to a viam carto object.
+func NewViamCarto(vcl CViamCartoLib) (CViamCarto, error) {
```

Comment:
Lets have this function actually take the viam-cartographer config & convert it into the appropriate C type